File: api/API-RetiScan/models/Analysis.js
const pool = require('../config/database');

const Analysis = {
    /**
     * Crea un nuevo análisis en estado PENDING.
     * @param {string} patientId
     * @param {string} doctorId   - Para aislamiento multi-tenant
     * @param {string} eye        - 'LEFT' o 'RIGHT'
     * @param {string} [imageUri]
     * @param {string} [doctorNotes]
     */
    async create(patientId, doctorId, eye, imageUri, doctorNotes = null) {
        const uri = imageUri || `retiscan://pending/${require('crypto').randomUUID()}`;
        const result = await pool.query(
            `INSERT INTO analyses (patient_id, doctor_id, eye, image_uri, doctor_notes, status, ai_result)
       VALUES ($1, $2, $3, $4, $5, 'PENDING', NULL)
       RETURNING *`,
            [patientId, doctorId, eye, uri, doctorNotes]
        );
        return result.rows[0];
    },

    /**
     * Recupera todos los análisis de un médico (aislamiento).
     * @param {string} doctorId
     */
    async findAllByDoctor(doctorId) {
        const result = await pool.query(
            'SELECT * FROM analyses WHERE doctor_id = $1 ORDER BY created_at DESC',
            [doctorId]
        );
        return result.rows;
    },

    /**
     * Encuentra todos los análisis de un paciente, verificando propiedad del médico.
     * @param {string} patientId
     * @param {string} doctorId
     */
    async findByPatientAndDoctor(patientId, doctorId) {
        const result = await pool.query(
            'SELECT * FROM analyses WHERE patient_id = $1 AND doctor_id = $2 ORDER BY created_at DESC',
            [patientId, doctorId]
        );
        return result.rows;
    },

    /**
     * Encuentra un análisis por UUID verificando propiedad del médico.
     * @param {string} id
     * @param {string} doctorId
     */
    async findByIdAndDoctor(id, doctorId) {
        const result = await pool.query(
            'SELECT * FROM analyses WHERE id = $1 AND doctor_id = $2',
            [id, doctorId]
        );
        return result.rows[0] || null;
    },

    /**
     * Variante para pacientes: encuentra análisis del paciente sin restricción de doctor.
     * @param {string} patientId
     */
    async findByPatientId(patientId) {
        const result = await pool.query(
            'SELECT * FROM analyses WHERE patient_id = $1 ORDER BY created_at DESC',
            [patientId]
        );
        return result.rows;
    },

    /** Encuentra un análisis por UUID (uso interno del worker de IA). */
    async findById(id) {
        const result = await pool.query(
            'SELECT * FROM analyses WHERE id = $1',
            [id]
        );
        return result.rows[0] || null;
    },

    /**
     * Actualiza el estado y opcionalmente el ai_result de un análisis.
     * @param {string} id
     * @param {'PENDING'|'PROCESSING'|'COMPLETED'|'FAILED'} status
     * @param {object|null} aiResult - Payload JSONB del modelo de IA
     */
    async updateStatus(id, status, aiResult = null) {
        const result = await pool.query(
            `UPDATE analyses
       SET status     = $1,
           ai_result  = $2,
           updated_at = NOW()
       WHERE id = $3
       RETURNING *`,
            [status, aiResult ? JSON.stringify(aiResult) : null, id]
        );
        return result.rows[0] || null;
    },

    /**
     * Elimina permanentemente un análisis (también en cascada sus registros).
     * Solo si pertenece al médico.
     */
    async deleteByIdAndDoctor(id, doctorId) {
        const result = await pool.query(
            'DELETE FROM analyses WHERE id = $1 AND doctor_id = $2 RETURNING id',
            [id, doctorId]
        );
        return result.rows[0] || null;
    },
};

module.exports = Analysis;

File: app/PWA-RetiScan/lib/models/analysis.dart
class Analysis {
  final String id;
  final String patientId;
  final String status; // PENDING | PROCESSING | COMPLETED | FAILED
  final DateTime createdAt;
  final Map<String, dynamic>? aiResult;
  final String? doctorNotes;
  final String? imageUri;

  Analysis({
    required this.id,
    required this.patientId,
    required this.status,
    required this.createdAt,
    this.aiResult,
    this.doctorNotes,
    this.imageUri,
  });

  bool get isPending => status == 'PENDING';
  bool get isProcessing => status == 'PROCESSING';
  bool get isCompleted => status == 'COMPLETED';
  bool get isFailed => status == 'FAILED';
  bool get isFinished => isCompleted || isFailed;

  factory Analysis.fromJson(Map<String, dynamic> json) {
    return Analysis(
      id: json['id'] as String? ?? json['_id'] as String? ?? '',
      patientId: json['patient_id'] as String? ?? json['patientId'] as String? ?? '',
      status: json['status'] as String? ?? 'PENDING',
      createdAt: json['created_at'] != null 
          ? DateTime.tryParse(json['created_at'].toString()) ?? DateTime.now()
          : json['createdAt'] != null
              ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
              : DateTime.now(),
      aiResult: json['ai_result'] as Map<String, dynamic>? ?? json['aiResult'] as Map<String, dynamic>?,
      doctorNotes: json['doctor_notes'] as String? ?? json['doctorNotes'] as String?,
      imageUri: json['image_url'] as String? ?? json['imageUrl'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'patient_id': patientId,
        'status': status,
        'created_at': createdAt.toIso8601String(),
        'ai_result': aiResult,
        'doctor_notes': doctorNotes,
        'image_url': imageUri,
      };
}